### PR TITLE
Add MSCCL Support

### PR DESCRIPTION
**Pull request by yzygitzh**
This PR integrates initial MSCCL executor support into RCCL. It includes several parts:

MSCCL core implementation, including:
- Add major MSCCL executor functionalities, including algorithm parser, lifecycle manager, context/connection initializer and algorithm executor kernel. The internal interfaces of these are defined in `src/include/msccl*`. Host code implementation is in `src/misc/msccl*`, and device code implementation is in `src/collectives/device/msccl*`.
- Add public MSCCL executor interface for MSCCL scheduler future usage, in src/nccl.h.in and src/collectives/msccl.cc.

Revisions to original RCCL code base, including:
- Add some new primitives required by MSCCL to RCCL's original primitive set, in src/collectives/device/prims*.
- Add calls to MSCCL executor lifecycle APIs in init.cc.
- Add calls to placeholder MSCCL scheduler in collective APIs, in src/collectives/*.cc.
- Export some internal interface in proxy for MSCCL usage, in src/proxy.cc and src/include/proxy.h.
- Disable shared buffer in net for MSCCL usage, in src/net.cc.
- Revision to make scripts.

Unit tests, including:
- Ring all-reduce algorithm with LL/LL128/Simple protocol in MSCCL format for UT, in tools/msccl-unit-test-algorithms/*.
- Add unit test for MSCCL, in test/AllReduce_Msccl.cpp. To run this test, use the following command in a 16-GPU system: RCCL_MSCCL_ENABLE=1 MSCCL_ALGO_FILE_PATH=../tools/msccl-unit-test-algorithms/all-reduce-ring-simple-n1g16.xml ./test/UnitTests --gtest_filter="AllReduce.Msccl*".

### CODE REVIEW

**Comment by akolliasAMD** [comment]
on the tests side, RunSimpleSweep now requires one extra parameter:
    testBed.RunSimpleSweep(funcTypes, dataTypes, redOps, roots, numElements,
                           inPlaceList, managedMemList, **useHipGraphList**);
can be initialized as:
std::vector<bool>           const useHipGraphList = {false, true};

**Comment by yzygitzh** [comment]
> on the tests side, RunSimpleSweep now requires one extra parameter: testBed.RunSimpleSweep(funcTypes, dataTypes, redOps, roots, numElements, inPlaceList, managedMemList, **useHipGraphList**); can be initialized as: std::vector const useHipGraphList = {false, true};

Thanks for reminding! Will update PR to fix this.

File: src/misc/msccl/msccl_lifecycle.cc [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+}
+
+bool mscclAvailable() {
+  return mscclEnabled() && mscclInitialized.load(std::memory_order_acquire);
+}
+
+ncclResult_t mscclInit(ncclComm_t comm) {
+  if (comm->intraRanks > 1) {
+    mscclInitialized.store(false, std::memory_order_release);
+    WARN("MSCCL doesn't support multiple GPUs in one process and is not available");
```
Comment (wenkaidu) [inline src/misc/msccl/msccl_lifecycle.cc:46]: Can we change this to INFO?

Comment (yzygitzh) [reply src/misc/msccl/msccl_lifecycle.cc]: Fixed.

**Review by akolliasAMD — APPROVED** [review]

**Review by wenkaidu — APPROVED** [review]

**Pull request merged**